packages/flutter/lib/src/widgets/scrollable_repaint_boundary.dart

```diff
@@ -0,0 +1,148 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'notification_listener.dart';
+import 'scroll_notification.dart';
+
+/// A [ScrollableRepaintBoundaryManager] tracks the scrollable state and notifies
+/// descendant [ScrollableRepaintBoundary].
+class ScrollableRepaintBoundaryManager extends StatefulWidget {
+  /// Create a new [ScrollableRepaintBoundaryManager].
+  const ScrollableRepaintBoundaryManager({required this.child, super.key});
+
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  @override
+  State<ScrollableRepaintBoundaryManager> createState() {
+    return _ScrollableRepaintBoundaryManager();
+  }
+}
+
+class _ScrollableRepaintBoundaryManager extends State<ScrollableRepaintBoundaryManager> {
+  final ValueNotifier<bool> boundary = ValueNotifier<bool>(false);
+
+  bool _onScrollUpdateNotification(ScrollNotification notification) {
+    if (notification.depth != 0) {
+      return false;
+    }
+    if (notification is ScrollStartNotification) {
+      boundary.value = true;
+    } else if (notification is ScrollUpdateNotification) {
+      boundary.value = true;
+    } else if (notification is ScrollEndNotification) {
+      boundary.value = false;
+    }
+    return false;
+  }
```

Comment:
This actually attempts to disable RPBs when scrolling stops - which is contrary to the stated behavior of this PR (sorry, I started this a long time ago).

That said, either strategy could be effective...

Reply:
Won't this make layer diffing/caching more expensive on the engine side?

Alternative question: won't this make the initial frames of scrolling slower?

I think there's probably a way we can do this by determining whether the scroll extent is > 0 (or if the physics is NeverScrollablePhysics) and just never adding them if scrolling will never happen.